src/core/Akka.Remote/Endpoint.cs

```diff
@@ -411,147 +414,174 @@ protected override void PostRestart(Exception reason)
+                    GoToActive();
+                }
+                else
                 {
-                    stop.ReplyTo.Tell(new EndpointWriter.StoppedReading(stop.Writer));
-                    Sender.Tell(new EndpointWriter.StoppedReading(stop.Writer));
-                });
+                    GoToIdle();
+                }
+            });
+            Receive<AttemptSysMsgRedelivery>(redelivery => { }); // Ignore
+            Receive<EndpointManager.Send>(send => send.Message is ISystemMessage, send => TryBuffer(send.Copy(NextSeq())));
+            Receive<EndpointManager.Send>(send => Context.System.DeadLetters.Tell(send));
+            Receive<EndpointWriter.FlushAndStop>(flush => Context.Stop(Self));
+            Receive<EndpointWriter.StopReading>(stop =>
+            {
+                stop.ReplyTo.Tell(new EndpointWriter.StoppedReading(stop.Writer));
+                Sender.Tell(new EndpointWriter.StoppedReading(stop.Writer));
+            });
         }
 
-        protected void Idle(object message)
+        protected void IdleBehavior()
         {
-            message.Match()
-                .With<EndpointManager.Send>(send =>
-                {
-                    _writer = CreateWriter();
-                    //Resending will be triggered by the incoming GotUid message after the connection finished
-                    HandleSend(send);
-                    Context.Become(OnReceive);
-                })
-                .With<AttemptSysMsgRedelivery>(sys =>
+            Receive<IsIdle>(idle => Sender.Tell(Idle.Instance));
+            Receive<EndpointManager.Send>(send =>
+            {
+                _writer = CreateWriter();
+                //Resending will be triggered by the incoming GotUid message after the connection finished
+                HandleSend(send);
+                GoToActive();
+            });
+
+            Receive<AttemptSysMsgRedelivery>(sys =>
+            {
+                if (_resendBuffer.Nacked.Any() || _resendBuffer.NonAcked.Any())
                 {
                     _writer = CreateWriter();
                     //Resending will be triggered by the incoming GotUid message after the connection finished
-                    Context.Become(OnReceive);
-                })
-                .With<EndpointWriter.FlushAndStop>(stop => Context.Stop(Self))
-                .With<EndpointWriter.StopReading>(stop => stop.ReplyTo.Tell(new EndpointWriter.StoppedReading(stop.Writer)));
+                    GoToActive();
+                }
+            });
+            Receive<TooLongIdle>(idle =>
+            {
+                HandleTooLongIdle();
+            });
+            Receive<EndpointWriter.FlushAndStop>(stop => Context.Stop(Self));
+            Receive<EndpointWriter.StopReading>(stop => stop.ReplyTo.Tell(new EndpointWriter.StoppedReading(stop.Writer)));
         }
 
-
+        protected void FlushWait()
```

Comment:
We enter the `FlushWait` behavior on the `ReliableDeliverySupervisor` when we're in the `Receiving` behavior and receive a `FlushAndStop` message. We forward that message onto the `EndpointWriter` and wait for it to terminate once it has emptied its buffer of messages. Upon receiving the `Terminated` message from the `EndpointWriter` we shut ourselves down also.